### PR TITLE
images switch server to centos:stream9 as base image

### PR DESCRIPTION
**Pull request by obnoxxx**
Fixes: #78

Signed-off-by: Michael Adam <obnox@samba.org>

### CODE REVIEW

**Comment by obnoxxx** [comment]
@phlogistonjohn  - I wanted to take a stab at fixin an issue... building  strangely fails locally for me. let's see how the ci goes ...

**Comment by obnoxxx** [comment]
ok. both server and ad-server builds fail in the CI here as well. will try to fix it later / tomorrow ...

**Comment by phlogistonjohn** [comment]
AFAIK, we discussed this a while back and one of the major blockers was that centos samba builds don't include the DC component at all. Thoughts?

**Comment by phlogistonjohn** [comment]
Just to install all the distro packages we need for the server image we'll need a bunch of commands like:
```
dnf --setopt=install_weak_deps=False install -y 'dnf-command(config-manager)'
dnf --setopt=install_weak_deps=False config-manager --set-enabled crb  highavailability resilientstorage
dnf --setopt=install_weak_deps=False install -y epel-release
dnf --setopt=install_weak_deps=False config-manager --set-enabled epel
```
In the install-packages.sh script.

However, you'll then hit the issue that sambacc fails to install because the rpm is built using fedora which is using python3.10 and when it tries to install sambacc it'll fail because the version of python is too low on cs9.

That's all on top of the issue of not having samba-dc rpm at all as mentioned earlier, so while I don't think making server cs9 based is insurmountable, there are a bunch of issues that need to be discussed IMO.

**Comment by anoopcs9** [comment]
@obnoxxx @phlogistonjohn 

I have the required changes still on hold for quite some time(as mentioned in https://github.com/samba-in-kubernetes/samba-container/issues/78#issuecomment-1246635766). I've rebased those changes against latest master for both sambacc and samba-container.

* https://github.com/samba-in-kubernetes/sambacc/commit/22faf7abb04896f11d59ca78774933e9ec26a326

* https://github.com/samba-in-kubernetes/samba-container/commit/075b6550528082358a44313bb6764932817521d5

**Comment by obnoxxx** [comment]
Updated the PR including the patch https://github.com/anoopcs9/samba-container/commit/075b6550528082358a44313bb6764932817521d5 from @anoopcs9   for the server image and trying to adjust for ad-server. 
It failed locally for me. let's see what the ci thinks ... @anoopcs9 , @phlogistonjohn

**Comment by anoopcs9** [comment]
> Updated the PR including the patch [anoopcs9@075b655](https://github.com/anoopcs9/samba-container/commit/075b6550528082358a44313bb6764932817521d5) from @anoopcs9 for the server image and trying to adjust for ad-server. It failed locally for me. let's see what the ci thinks ... @anoopcs9 , @phlogistonjohn

You'll have to first make changes in sambacc as I mentioned earlier in https://github.com/samba-in-kubernetes/samba-container/pull/98#issuecomment-1396429224. Only then we get an el9 build of **python3-sambacc** for install during server/ad-server build process.

**Comment by phlogistonjohn** [comment]
@anoopcs9 can I suggest that if you have changes ready for sambacc you file a PR on that repo? I have some q's and thoughts but I think a PR (even just a draft PR) is the better way to discuss it. Thanks!

**Comment by obnoxxx** [comment]
@phlogistonjohn , @anoopcs9 , seems I have managed to fix all the server and ad-server builds now in the ci. 😄

File: images/ad-server/install-packages.sh [hunk 1]
```diff
@@ -17,22 +16,27 @@ case "${install_packages_from}" in
...
     python3-pyxattr \
-    "samba-dc${samba_version_suffix}" \
-    procps-ng \
-    /usr/bin/smbclient
+    "samba${samba_version_suffix}" \
+    "samba-client${samba_version_suffix}" \
+    "samba-winbind${samba_version_suffix}" \
+    "samba-winbind-clients${samba_version_suffix}" \
+    tdb-tools \
+    "ctdb${samba_version_suffix}"
```
Comment (phlogistonjohn) [inline images/ad-server/install-packages.sh:38]: It's possible I'm reading this wrong, but it appars to replace the samba-dc package with samba file server packages -- effectively turning the ad-server image into a copy of the (file) server image. I'm very confused.

Comment (obnoxxx) [reply images/ad-server/install-packages.sh:38]: @phlogistonjohn ,  I think I did the change initially by copying the files from @anoopcs9 's patch which in fact touched the file server image  😬  ...

**Comment by phlogistonjohn** [comment]
@mergifyio rebase

**Comment by obnoxxx** [comment]
rebasing locally worked without conflicts. git magically applied @anoopcs9 's original patch to the  now inappropriately named `Containerfile.fedora`

After a quick discussion with @phlogistonjohn , I decided to misuse this PR to generalize the multi-flvour build just introduced by opensuse builds in PR #104, ant I added commits to add explicit choices for fedora and centos builds. 

example build commands are now:

* `make -f Makefile.opensuse build-client `
*  `make -f Makefile.fedora build-client`
* `make -f Makefile.centos build-nightly-server`

We should probably add explicitci actions  for those ...

**Comment by obnoxxx** [comment]
fwiw, the default is still fedora.
and I forgot to mention that I dropped the commit to change the ad-server image. 

@anoopcs9 's original patch to switch server to centos probably needs amending to change rge default to centos as the commit claims to do ...

@phlogistonjohn  - what do you think?

**Comment by obnoxxx** [comment]
updated the PR to fix some typos and other issus. now some checks are failing. e.g.:
```
Run ./tests/test-samba-ad-server-kubernetes.sh
Creating ad server deployment...
NAME              READY   UP-TO-DATE   AVAILABLE   AGE
samba-ad-server   0/1     1            0           1s
Samba ad pod is samba-ad-server-6[8](https://github.com/samba-in-kubernetes/samba-container/actions/runs/4008915832/jobs/6883730540#step:9:9)fcf68d58-blkr8
waiting for pod to be in Running state
.
NAME                               READY   STATUS   RESTARTS   AGE
samba-ad-server-68fcf68d58-blkr8   0/1     Error    0          2s

waiting for samba to become reachable
........................................................................................................................
Error: samba ad did not become reachable
Error: Process completed with exit code 1.
```

**Comment by obnoxxx** [comment]
@phlogistonjohn  , @anoopcs9 , @synarete - what do you think  about this?

File: Makefile.fedora [hunk 2]
```diff
@@ -0,0 +1,19 @@
+include Makefile
```
Comment (phlogistonjohn) [inline Makefile.fedora:1]: I'm only 50/50 on adding a file like this as it is redundant with the defaults in Makefile. If we change something we now have to change it in two places. That said, it has the benefit of being consistent with the recent changes to add opensuse support.

File: Makefile.centos [hunk 3]
```diff
@@ -0,0 +1,19 @@
+include Makefile
+
+SERVER_SRC_FILE:=$(SERVER_DIR)/Containerfile.centos
+AD_SERVER_SRC_FILE:=$(AD_SERVER_DIR)/Containerfile.centos
```
Comment (phlogistonjohn) [inline Makefile.centos:4]: I'd rather not add any overrides that don't exist yet. So for example  images/ad-server/Containerfile.centos doesn't exist... so it's never going to do anything useful.
I'd be OK if you left some of these variable names in but commented them out, just to serve as breadcrumbs.

File: images/server/install-packages.sh [hunk 4]
```diff
@@ -6,7 +6,6 @@ get_custom_repo() {
     url="$1"
     fname="$(basename "$url")"
     dest="/etc/yum.repos.d/${fname}"
-    dnf install --setopt=install_weak_deps=False -y curl
```
Comment (phlogistonjohn) [inline images/server/install-packages.sh:9]: If we support building from either centos or fedora we should make the script work on both (or have two scripts). Right now, this change will simply cause the build to fail on fedora.

File: Makefile [hunk 5]
```diff
@@ -24,17 +24,17 @@ SERVER_DIR:=images/server
...
 	$(SERVER_DIR)/install-sambacc.sh
-AD_SERVER_SRC_FILE:=$(AD_SERVER_DIR)/Containerfile.fedora
+AD_SERVER_SRC_FILE:=$(AD_SERVER_DIR)/Containerfile.centos
 AD_SERVER_SOURCES:=\
 	$(AD_SERVER_DIR)/install-packages.sh \
 	$(AD_SERVER_DIR)/install-sambacc.sh
-CLIENT_SRC_FILE:=$(CLIENT_DIR)/Containerfile.fedora
-TOOLBOX_SRC_FILE:=$(TOOLBOX_DIR)/Containerfile.fedora
+CLIENT_SRC_FILE:=$(CLIENT_DIR)/Containerfile.centos
+TOOLBOX_SRC_FILE:=$(TOOLBOX_DIR)/Containerfile.centos
```
Comment (phlogistonjohn) [inline Makefile:37]: These files don't exist yet.

File: Makefile [hunk 6]
```diff
@@ -24,17 +24,17 @@ SERVER_DIR:=images/server
...
 CLIENT_DIR:=images/client
 TOOLBOX_DIR:=images/toolbox
-SERVER_SRC_FILE:=$(SERVER_DIR)/Containerfile.fedora
+SERVER_SRC_FILE:=$(SERVER_DIR)/Containerfile.centos
 SERVER_SOURCES:=\
 	$(SERVER_DIR)/smb.conf \
 	$(SERVER_DIR)/install-packages.sh \
 	$(SERVER_DIR)/install-sambacc.sh
-AD_SERVER_SRC_FILE:=$(AD_SERVER_DIR)/Containerfile.fedora
+AD_SERVER_SRC_FILE:=$(AD_SERVER_DIR)/Containerfile.centos
```
Comment (phlogistonjohn) [inline Makefile:32]: This file doesn't exist yet.

File: images/server/install-sambacc.sh [hunk 7]
```diff
@@ -34,7 +34,7 @@ case $action in
         container_json_file="/usr/local/share/sambacc/examples/minimal.json"
     ;;
     install-rpm)
-        dnf install -y "${rpmfiles[0]}"
+        dnf install --skip-broken --nobes -y "${rpmfiles[0]}"
```
Comment (phlogistonjohn) [inline images/server/install-sambacc.sh:37]: Adding `--skip-broken` makes me think were asking to miss important error cases in the future. Unless you (plural) convince me it is necessary.

Comment (obnoxxx) [reply images/server/install-sambacc.sh:37]: > Adding `--skip-broken` makes me think were asking to miss important error cases in the future. Unless you (plural) convince me it is necessary.
@anoopcs9  has to do the convincing ... It came from him.

File: Makefile (hunk 5 above)
Comment (obnoxxx) [reply Makefile]: I changed the commit to only add a server centos flavor ...

File: Makefile.centos (hunk 3 above)
Comment (obnoxxx) [reply Makefile.centos:4]: I changed the commit to only add the server centis build flavor, but it still adds overrides to files that don't exist (yet) ...  we could add those in the future if we want to.

File: Makefile (hunk 5 above)
Comment (obnoxxx) [reply Makefile]: Now I added a second commit to add the client centos flavor

**Comment by obnoxxx** [comment]
@anoopcs9 , I changed your original commit in the following way:

* split the modification of the install scripts out into an earlier commit that enablrs the explicit centos server build via Makefile.centos:
*  changed your commit to only change the main Makefile to default to centos for the server build.

Please check if you are ok with the approach in principle.

**Comment by obnoxxx** [comment]
@phlogistonjohn , @anoopcs9 , @synarete , @spuiuk : if desired, I could split the commits that add explicit build flavors for fedora and centos out into a separate PR (or several), since this does not match the  originaltopic of the PR any longer ...

File: Makefile (hunk 6 above)
Comment (obnoxxx) [reply Makefile]: > This file doesn't exist yet.

fixed by adding a commit before this.

**Comment by phlogistonjohn** [comment]
I would really like to break this PR up into smaller pieces. For example, the changes to install packages script will break fedora build, but this is masked in this PR because you also change the default behavior of the makefile.

Would you be open to creating or updating this PR to **only** **add** support for centos? That would mean a new Makefile.centos{9?}, a new Containerfile.centos9, and updates to the script that work for both fedora and centos.

**Comment by obnoxxx** [comment]
@phlogistonjohn , as discussed earlier today, I have split the generalisation part and supporting fedora and centos as explicit flavors out into PR #107 . This PR now depends on that other PR

**Comment by obnoxxx** [comment]
I don't know if we still want to do this. In our discussions, I had the impression, that you don't want to change the default base OS, @phlogistonjohn - right?

**Pull request closed without merging**